libs/src/services/discoveryProvider/DiscoveryProviderSelection.js

```diff
@@ -0,0 +1,213 @@
+  shortcircuit () {
+    return this.getCached()
+  }
+
+  async select () {
+    const endpoint = await super.select()
+    this.setCached(endpoint)
+    return endpoint
+  }
+
+  /**
+   * Checks whether a given response is healthy:
+   * - Not behind in blocks
+   * - 200 response
+   * - Current version
+   *
+   * Other responses are collected in `this.backups` if
+   * - Behind by only a patch version
+   *
+   * @param {Response} response axios response
+   * @param {{ [key: string]: string}} urlMap health check urls mapped to their cannonical url
+   * e.g. https://discoveryprovider.audius.co/health_check => https://discoveryprovider.audius.co
+   */
+  isHealthy (response, urlMap) {
+    const { status, data } = response
+    const { block_difference: blockDiff, service, version } = data
+    if (status !== 200) return false
+    if (service !== DISCOVERY_SERVICE_NAME) return false
+    if (!semver.valid(version)) return false
+    if (!this.ethContracts.isValidSPVersion(version, this.currentVersion)) return false
+
+    if (
+      blockDiff > UNHEALTHY_BLOCK_DIFF ||
+      version !== this.currentVersion
+    ) {
+      this.addBackup(urlMap[response.config.url], response.data)
+      return false
+    }
+
+    return true
+  }
+
+  /**
+   * Estabilishes that connection to discovery providers has regressed
+   */
+  enterRegressedMode () {
+    console.info('Entering regressed mode')
+    this._regressedMode = true
+    setTimeout(() => {
+      console.info('Leaving regressed mode')
+      this._regressedMode = false
+    }, REGRESSED_MODE_TIMEOUT)
+  }
+
+  isInRegressedMode () {
+    return this._regressedMode
+  }
+
+  /**
+   * In the case of no "healthy" services, we resort to backups.
+   * 1. Pick the most recent version that's not behind
+   * 2. Pick the least behind provider and enter "regressed mode"
+   */
+  async selectFromBackups () {
+    const versions = []
+    const blockDiffs = []
+
+    const versionMap = {}
+    const blockDiffMap = {}
+
+    // Go backwards in time on chain and get the registered versions up to PREVIOUS_VERSIONS_TO_CHECK.
+    // Record those versions in a set and validate any backups against that set.
+    // TODO: Clean up this logic when we can validate a specific version rather
+    // than traversing backwards through all the versions
+    if (!this.validVersions) {
+      this.validVersions = new Set([this.currentVersion])
+      const numberOfVersions = await this.ethContracts.getNumberOfVersions(DISCOVERY_SERVICE_NAME)
+      for (let i = 0; i < Math.min(PREVIOUS_VERSIONS_TO_CHECK, numberOfVersions - 1); ++i) {
+        const pastServiceVersion = await this.ethContracts.getVersion(
+          DISCOVERY_SERVICE_NAME,
+          // Exclude the latest version when querying older versions
+          // Latest index is numberOfVersions - 1, so 2nd oldest version starts at numberOfVersions - 2
+          numberOfVersions - 2 - i
```

Comment:
if we use `PREVIOUS_VERSIONS_TO_CHECK`, and `PREVIOUS_VERSIONS_TO_CHECK < numberOfVersions-1`, there may be a possibility that `numberOfVersions-2-i` may go out of bounds?

Reply:
I don't think so because iut's min(PREVIOUS_V_TO_CHECK, numVersions - 1)

If numVersions === 0 or 1, this loop won't run
If numVersions === 2 versions===([a, b]), we'll get numVersions - 2 which is versions[0]